### ISSUE TITLE
CosineEmbeddingLoss loss does not make sense

### ISSUE THREAD

**Issue by isalirezag** [pytorch/pytorch#14173]
I am not sure if it is a bug or im doing some mistake
shouldn't the out put of the following example be 1?

```
import torch
import torch.nn.functional as F
from torch.autograd import Variable
input1 = torch.rand(1,2,10,10)

y = torch.ones(1,1,10,10)
y.requires_grad = False

loss = torch.nn.CosineEmbeddingLoss()
output = loss(input1,input1,y)
print(output)
Variable containing:
1.00000e-07 *
  4.1723
[torch.FloatTensor of size 1]

```

Im using pytorch 0.3.0 and ubuntu 16.04

**Comment by vishwakftw** [comment]
Please have a look at the definition [here](https://pytorch.org/docs/master/nn.html#torch.nn.CosineEmbeddingLoss). Cosine between two equal vectors is 1, so the answer is correct to machine precision.

**Comment by isalirezag** [comment]
yeah but it still does not make sense, so the output should be 0 here right?
why it is not absolutely zero?

**Comment by soumith** [comment]
@isalirezag because floating point precision doesn't return perfect numbers.

**Comment by gauravkoradiya** [comment]
you are right @vishwakftw  .... your output should be zero.
